docs/DesignDocs/ReproducibleBuildUsingNix.md

```diff
@@ -0,0 +1,397 @@
+  cache to provide prebuilding, or everything would require everything be built from source..
+
+### Caching and Hydra
+
+The official branches in nixpkgs are prebuilt and available at cache.nixos.org. Caching is keyed to the build outputs
+sha digest, so custom environments will be able to use those prebuilts so long as no changes have been made. 
+If any changes are made to that package, any package related to the changed package will need to be custom
+built and cached in nix store.
+
+
+## A Nix build system for OpenEnclave SDK
+
+We propose to add an optional nix based build environment for OpenEnclave SDK using docker containers.
+
+This would consist of :
+- Dockerfile.nix and build-oe-nix-build.sh for building the oe nix docker image locally.
+- build_openenclave.sh for building the oe deb package via a nix environment.
+- Dockerfile.test and test-oe-nix-build.sh for verifying that package is functional.
+
+The process builds a reproducible .deb package of openenclave-sdk. 
+
+This process can be integrated with ADO or CPDX CI, but that is not part of this submission. We do not currently
+recommend depending exclusively on a nix environment for production packaging and CI.
+
+### Dockerfile.nix
+
+The dockerfile is passed up to four build args:
+
+ - BASE_IMAGE allows for different os base images, in particular it allows the substitution of a
+   CPDX base image rather than the generic ubuntu 20.04. As was previously mentioned, we are basing on 
+   ubuntu 20.04 because package versions align best with nixpkgs release-20.03.
+
+ - BUILD_USER, BUILD_USER_ID, BUILD_USER_HOME need to be set together.
+   Reproducibility requires a standard build user, since the tar directory entries in the deb package include 
+   user and group information.  We default to "azureuser" "1000" "/home/azureuser"
+
+The script "build-oe-nix-image.sh" is a thin wrapper. It takes no arguments, and just passes the needed arguments
+to docker build.  If it needs to be modified, a new script should be created.
+
+Dockerfile.nix prepares the build image by:
+
+- adding via apt minimum packages for the outer image. This includes a text editor (vim) and ELF patch utility (patchelf).
+
+- sets up nit configuration.
+
+- diables apt.
+
+- creates the build user and directories, groups etc.
+
+- creates a nix expression for building openenclave via (shell.nix).
+
+- installs nix as the build user
+
+- calls the prep-nix-build.sh script which sets up the git clone of nix pkgs and preinstalls the package
+  derivations needed for building openenclave into the nix store. The search process for dependencies in
+  nix is to refer to the nix store first, before looking at nixpkgs or contacting the cache if enabled. While
+  we cannot select package versions from the expression, but must depend on the nixpkgs repo to provide the
+  correct versions, we can prime the nix store to provide the exact derivation we wish before referring to the repo.
```

Comment:
Intel SGX SDK [reproducible build](https://github.com/intel/linux-sgx/blob/544159a94a59547a18b7bff1d28a94bb14053572/linux/reproducibility/Dockerfile) seems to relying on specific version of packages in nixpkgs repo/cache directly. What's the benefit of the "prime the nix store" process?

Reply:
We lock the .drvs into the nix store in the docker build process. This preserves the relationship between the package and its dependencies, as well as preventing garbage collection of the drvs until the correct packages are installed.

Reply:
Note the nix-env which retrieves or builds the build outputs from the .drv uses the expected sha digest of the packages, which will alert us if something else is coming up.

Reply:
@yakman2020 , what does the term "packages" in  "This preserves the relationship between the package and its dependencies, as well as preventing garbage collection of the drvs until the correct packages are installed."  refer to? Does it refer to `nix-build` output, as the "qsgajw2yhrf33qgibsvd9y8zaw64w7j7-helloWorld" in the example? If that's the case, once the pairing between "znmj0axqf4v5mjwswd9bq3dvxf0jcsah-test-environment.drv" and "qsgajw2yhrf33qgibsvd9y8zaw64w7j7-helloWorld" is established and recorded for audit purpose, can the docker image directly use the "packages" in nixpkgs repo/cache and bypass the "prime the nix store" process?

Reply:
nix always uses any realised build output in nix store if it has them.  nix-env given a drv gets them from the prebuilt cache/channel.  But ther eis no auditability past that.  The drv points at the build output, the build output does not point at the drv.  A lone build output in nix is just as anonymous as a deb package.